### PR TITLE
update vendored libsecp to 0.3.2

### LINKED ISSUE #631: Update secp256k1 C dependency to 0.3.1
The currently included version is 0.2 as far as I can see.  Since then two potential sidechannel vulnerabilities have been fixed (see [releases](https://github.com/bitcoin-core/secp256k1/releases).

### PR DESCRIPTION
**Pull request by apoelstra**
Fixes #631.

Also bumps the major version of rust-secp256k1-sys since this bump involves renaming all the symbols, so we might as well do it now. But we won't release after this PR; will wait until we're sure we're ready to do the release.

To reproduce run

```
./vendor-libsecp.sh acf5c55ae6a94e5ca847e07def40427547876101
```

### CODE REVIEW

**Comment by apoelstra** [comment]
CI failure is because we have a `test_raw_ctx` which is double-dropping context objects. In #261 where we introduced the double-drop I claimed that the first one "was a no-op, which is the point of the test". I need to investigate whether this used to be true and when it stopped being true..

**Comment by apoelstra** [comment]
Should be good to go. This wound up being more involved than I expected. Upstream changed the semantics of `secp256k1_preallocated_context_destroy` so that you can't call it twice in a row anymore which exposed some weirdness in our unit tests.

File: contrib/_test.sh [hunk 1]
```diff
@@ -23,7 +23,7 @@ fi
 # Test if panic in C code aborts the process (either with a real panic or with SIGILL)
 cargo test -- --ignored --exact 'tests::test_panic_raw_ctx_should_terminate_abnormally' 2>&1 \
     | tee /dev/stderr \
-    | grep "SIGILL\\|\[libsecp256k1] illegal argument. !rustsecp256k1_v0_._._fe_is_zero(&ge->x)"
+    | grep "SIGILL\\|\[libsecp256k1] illegal argument. "
```
Comment (tcharding) [inline contrib/_test.sh:26]: Why this change? It works still, right? You just don't like the grep statement?

File: src/lib.rs [hunk 2]
```diff
@@ -612,14 +586,35 @@ mod tests {
         assert!(vrfy.verify_ecdsa(&msg, &sig, &pk).is_ok());
         assert!(full.verify_ecdsa(&msg, &sig, &pk).is_ok());
 
+        // The following drop will have no effect; in fact, they will trigger a compiler
+        // error because manually dropping a `ManuallyDrop` is almost certainly incorrect.
```
Comment (tcharding) [inline src/lib.rs:590]: Did you mean to write "will trigger a clippy warning"?

**Review by tcharding — COMMENTED** [review]
Light review, I didn't verify the changes to `depend`, assuming they are done mechanically. Also I did not fully grok the `ManuallyDrop` unit test changes. I don't claim to be an unsafe code reviewer at the moment and did not allocate clock cycles to understanding this particular thing fully.

File: contrib/_test.sh (hunk 1 above)
Comment (apoelstra) [reply contrib/_test.sh:26]: No, the assertion has changed in the latest version of libsecp.

File: src/lib.rs (hunk 2 above)
Comment (apoelstra) [reply src/lib.rs:590]: No, in 1.72 and higher it will be a compiler warning.

Comment (tcharding) [reply src/lib.rs:590]: Oh, my bad.

File: contrib/_test.sh (hunk 1 above)
Comment (tcharding) [reply contrib/_test.sh:26]: Oh, cool. Thanks.

**Review by tcharding — APPROVED** [review]
ACK

**Comment by tcharding** [comment]
We want  to get this in before release, right?

**Comment by sanket1729** [comment]
Can you also add how to reproduce this in the commit message? Would help me review. I think the vendoring scripts now have some env variables that I need to set to reproduce.

**Comment by apoelstra** [comment]
@sanket1729 put a command in the commit description.

**Comment by tcharding** [comment]
Gentle bump please crew.

**Comment by tcharding** [comment]
Do we want to jump straight to v0.4.0?

**Comment by Davidson-Souza** [comment]
+1 on bumping to v0.4.0. Since this release is the only one with the EllSwift module, needed for #627.

**Comment by tcharding** [comment]
Yep no sweat I can do it.

**Comment by tcharding** [comment]
I just realized that releasing with libsecp 0.4.0 is not trivial because we need to write the wrappers for the new EllSwift stuff, given that (and the fact that I can't satisfy the linker in #652) as well as the fact that we are behind schedule on the rust-bitcoin release can we merge and release using this PR and then do libsecp 0.4.0 upgrade straight after release?

**Comment by Davidson-Souza** [comment]
@tcharding 
> we need to write the wrappers for the new EllSwift stuff

Technically, the code exists (#627). But it needs to be reviewed, and may require changes after #652

**Comment by tcharding** [comment]
Oh my bad, I did not realise. Will have another go, thanks.

**Comment by tcharding** [comment]
We can close this one now.
